hail/src/main/scala/is/hail/sparkextras/ContextRDD.scala

```diff
@@ -5,8 +5,48 @@ import org.apache.spark._
 import org.apache.spark.rdd._
 import org.apache.spark.ExposedUtils
 
+import scala.collection.JavaConverters._
 import scala.reflect.ClassTag
 
+class AssociativeCombiner[U](zero: U, combine: (U, U) => U) {
+  case class TreeValue(var value: U, var end: Int)
+
+  private val t = new java.util.TreeMap[Int, TreeValue]()
+
+  def combine(i: Int, value0: U) {
+    var value = value0
+    var end = i
+
+    val nexttv = t.get(i + 1)
+    if (nexttv != null) {
```

Comment:
Is there a reason you use `floorEntry` below and not `ceilingEntry` here? I would have thought these two pieces of code would be mostly symmetrical.

Reply:
The reason it is asymmetric is because the key is the start of the interval and the end is the value, and in one case we want to compare to the start and the other we want to compare to the end.

We're inserting `i`.  It might merge with an interval on the left (if its end == i - 1) or the right (if its  start == i + 1), or both.

So we can directly look up i + 1, but we need floor to find the closest interval to the left (and then check that its end is i - 1).

Make sense?

Reply:
got it!